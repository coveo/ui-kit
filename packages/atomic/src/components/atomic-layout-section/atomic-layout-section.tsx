import {Component, Prop} from '@stencil/core';
import {Section} from './sections';

@Component({
  tag: 'atomic-layout-section',
  shadow: false,
})
export class AtomicLayoutSection {
  @Prop() public section!: Section;
  @Prop() public minWidth?: string;
  @Prop() public maxWidth?: string;
}
