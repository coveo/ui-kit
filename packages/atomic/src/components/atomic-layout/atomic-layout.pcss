@import './search-layout.pcss';

atomic-layout {
  width: 100%;
  height: 100%;
  display: grid;
}

atomic-layout-section[section='search'] {
  grid-area: atomic-section-search;
}

atomic-layout-section[section='facets'] {
  grid-area: atomic-section-facets;
}

atomic-layout-section[section='main'] {
  grid-area: atomic-section-main;
}

atomic-layout-section[section='status'] {
  grid-area: atomic-section-status;
}

atomic-layout-section[section='pagination'] {
  grid-area: atomic-section-pagination;
}
