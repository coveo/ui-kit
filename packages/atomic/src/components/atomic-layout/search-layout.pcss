atomic-layout[layout='search'] {
  grid-template-areas:
    'atomic-section-search'
    'atomic-section-main';

  atomic-layout-section[section='search'] {
    margin: var(--atomic-layout-spacing-y) 0;
  }

  atomic-layout-section[section='facets'] {
    display: none;

    & * {
      margin-bottom: var(--atomic-layout-spacing-y);
    }
  }

  atomic-layout-section[section='main'] {
    margin-bottom: var(--atomic-layout-spacing-y);
  }

  atomic-layout-section[section='status'] {
    display: grid;
    justify-content: space-between;
    grid-template-areas:
      'atomic-breadbox      atomic-breadbox'
      'atomic-query-summary atomic-sort'
      'atomic-did-you-mean  atomic-did-you-mean';

    & > * {
      margin-bottom: var(--atomic-layout-spacing-y);
    }

    atomic-breadbox {
      grid-area: atomic-breadbox;
    }

    atomic-query-summary {
      grid-area: atomic-query-summary;
      align-self: center;
      overflow: hidden;
      min-width: 180px;
    }

    atomic-sort-dropdown {
      grid-area: atomic-sort;
      justify-self: end;
    }

    atomic-refine-toggle {
      grid-area: atomic-sort;
    }

    atomic-did-you-mean {
      grid-area: atomic-did-you-mean;
    }
  }

  atomic-layout-section[section='pagination'] {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;

    atomic-load-more-results {
      width: 100%;
    }

    & > * {
      margin-top: var(--atomic-layout-spacing-y);
    }

    /* Approx. width of pager & results per page */
    @media only screen and (min-width: 800px) {
      flex-direction: row;
    }
  }
}
