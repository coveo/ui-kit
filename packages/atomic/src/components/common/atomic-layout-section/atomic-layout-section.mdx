import {Canvas, Markdown, ArgTypes, Meta, Title, Description, Controls, Primary, Stories, Subtitle} from '@storybook/blocks';
import * as AtomicLayoutSectionStories from './atomic-layout-section.new.stories';

<Meta of={AtomicLayoutSectionStories} />

<Title />
<Subtitle ><Description/></Subtitle>

<Canvas
  additionalActions={[
    {
      title: 'Open in GitHub',
      onClick: () => {
        window.open(
          'https://github.com/coveo/ui-kit/blob/master/packages/atomic/src/components/common/atomic-layout-section/atomic-layout-section.ts',
          '_blank'
        );
      },
    },
  ]}
  of={AtomicLayoutSectionStories.Default}
/>

## Usage
This component is used to define a section within a layout. 
Each `atomic-layout-section` reserves space based on its section attribute, 
allowing the layout to position and style content consistently within the `atomic-search-layout` or `atomic-commerce-layout`.

An example using the `atomic-commerce-interface` and `atomic-commerce-layout` where we define sections for **search**, **main** content, **status**, **products**, and **pagination**:

```html
<atomic-commerce-interface>

  <atomic-commerce-layout>

    <atomic-layout-section section="search">
        ...
    </atomic-layout-section>
    
    <atomic-layout-section section="main">
        
      <atomic-layout-section section="status">
       ... 
      </atomic-layout-section>
      
      <atomic-layout-section section="products">
       ... 
      </atomic-layout-section>

      <atomic-layout-section section="pagination">
       ... 
      </atomic-layout-section>

    </atomic-layout-section>
  </atomic-commerce-layout>
</atomic-commerce-interface>
```

## Examples
<Stories title={""}/>

## Reference
<ArgTypes />