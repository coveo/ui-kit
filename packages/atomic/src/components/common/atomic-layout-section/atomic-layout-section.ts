import {LitElement} from 'lit';
import {customElement, property} from 'lit/decorators.js';
import {Section} from './sections';

/**
 * The `atomic-layout-section` lets you identify various sections for the related `atomic-search-layout` or `atomic-commerce-layout` component.
 *
 * @slot default - The default slot where you can add child components to the section.
 */
@customElement('atomic-layout-section')
export class AtomicLayoutSection extends LitElement {
  /**
   * The name of the layout section.
   */
  @property({type: String, reflect: true}) section!: Section;

  /**
   * For column sections, the minimum horizontal space it should take.
   * E.g. '300px'
   */
  @property({type: String, reflect: true, attribute: 'min-width'})
  minWidth?: string;
  /**
   * For column sections, the maximum horizontal space it should take.
   * E.g. '300px'
   */
  @property({type: String, reflect: true, attribute: 'max-width'})
  maxWidth?: string;

  createRenderRoot() {
    return this;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'atomic-layout-section': AtomicLayoutSection;
  }
}
