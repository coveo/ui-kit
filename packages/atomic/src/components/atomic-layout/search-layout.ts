import {findSection, sectionSelector} from '../atomic-layout-section/sections';

export function buildSearchLayout(
  element: HTMLElement,
  mobileBreakpoint: string
) {
  const id = element.id;
  const layoutSelector = `atomic-search-layout#${id}`;
  const mediaQuerySelector = `@media only screen and (min-width: ${mobileBreakpoint})`;

  const displayLayout = `${layoutSelector} { display: grid }`;

  const facetsDeclarations = () => {
    const facetsSection = findSection(element, 'facets');
    const mainSection = findSection(element, 'main');
    if (!facetsSection || !mainSection) {
      return '';
    }

    const facetsMin = facetsSection.minWidth || '272px';
    const facetsMax = facetsSection.maxWidth || '350px';
    const mainMin = mainSection.minWidth || '50%';
    const mainMax = mainSection.maxWidth || '1100px';

    return `${mediaQuerySelector} {
      ${layoutSelector} {
        grid-template-areas:
          '. .                     atomic-section-search .'
          '. atomic-section-facets atomic-section-main   .'
          '. atomic-section-facets .                     .';
        grid-template-columns: 
          1fr minmax(${facetsMin}, ${facetsMax}) minmax(${mainMin}, ${mainMax}) 1fr;
        column-gap: var(--atomic-layout-spacing-x);
      }

      ${layoutSelector} ${sectionSelector('facets')} {
        display: block;
      }
    }`;
  };

  const refineDeclarations = () => {
    const statusSection = findSection(element, 'status');
    if (!statusSection) {
      return '';
    }

    const refineToggle = statusSection.querySelector('atomic-refine-toggle');
    if (!refineToggle) {
      return '';
    }

    const statusSelector = `${layoutSelector} ${sectionSelector('status')}`;
    return `${statusSelector} atomic-sort-dropdown {
      display:none;
    }

    ${mediaQuerySelector} {
     ${statusSelector} atomic-sort-dropdown {
       display:block;
      }

      ${statusSelector} atomic-refine-toggle {
        display: none;
       }
    }`;
  };

  return [displayLayout, facetsDeclarations(), refineDeclarations()].join('');
}
