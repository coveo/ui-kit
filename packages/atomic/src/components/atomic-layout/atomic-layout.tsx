import {Component, Element, Prop} from '@stencil/core';
import {randomID} from '../../utils/utils';
import {buildSearchLayout} from './search-layout';

export type Layout = 'search';

/**
 * The `atomic-layout` helps organize elements in the page relative to one another.
 */
@Component({
  tag: 'atomic-layout',
  styleUrl: 'atomic-layout.pcss',
  shadow: false,
})
export class AtomicLayout {
  @Element() private host!: HTMLElement;
  /**
   * Name of the layout.
   */
  @Prop({reflect: true}) public layout!: Layout;
  // TODO: make breakpoint configurable & reuse in templates
  private breakpoint = '1024px';

  public connectedCallback() {
    const id = this.host.id || randomID('atomic-layout-');
    this.host.id = id;

    const styleTag = document.createElement('style');
    styleTag.innerHTML = this.style;
    this.host.appendChild(styleTag);
  }

  private get style() {
    switch (this.layout) {
      case 'search':
        return buildSearchLayout(this.host, this.breakpoint);
    }
  }
}
