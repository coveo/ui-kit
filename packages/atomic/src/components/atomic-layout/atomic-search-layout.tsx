import {Component, Element, Prop} from '@stencil/core';
import {randomID} from '../../utils/utils';
import {buildSearchLayout} from './search-layout';

/**
 * The `atomic-search-layout` helps organize elements in the page.
 */
@Component({
  tag: 'atomic-search-layout',
  styleUrl: 'atomic-search-layout.pcss',
  shadow: false,
})
export class AtomicSearchLayout {
  @Element() private host!: HTMLElement;
  /**
   * CSS valuethat defines where the layout goes from mobile to desktop.
   * e.g., 800px, 65rem.
   */
  @Prop({reflect: true}) public mobileBreakpoint = '1024px';

  public connectedCallback() {
    const id = this.host.id || randomID('atomic-search-layout-');
    this.host.id = id;

    const styleTag = document.createElement('style');
    styleTag.innerHTML = buildSearchLayout(this.host, this.mobileBreakpoint);
    this.host.appendChild(styleTag);
  }
}
