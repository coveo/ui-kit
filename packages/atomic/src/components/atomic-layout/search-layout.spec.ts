import {buildSearchLayout} from './search-layout';
import {newSpecPage} from '@stencil/core/testing';
import {AtomicLayoutSection} from '../atomic-layout-section/atomic-layout-section';
import {AtomicLayout} from './atomic-layout';

describe('Search Layout', () => {
  describe('#buildSearchLayout', () => {
    async function getSearchLayout(html: string) {
      const page = await newSpecPage({
        components: [AtomicLayoutSection, AtomicLayout],
        html,
      });
      return buildSearchLayout(
        page.body.querySelector('atomic-layout')!,
        '1024px'
      );
    }

    it(`when the atomic-layout component is empty
    should match snapshot`, async () => {
      const result = await getSearchLayout(
        '<atomic-layout id="abc" layout="search"></atomic-layout>'
      );
      expect(result).toMatchSnapshot();
    });

    it(`when the atomic-layout component contains a main & facets sections
    should match snapshot`, async () => {
      const result =
        await getSearchLayout(`<atomic-layout id="abc" layout="search">
            <atomic-layout-section section="facets" min-width="200px" max-width="300px"></atomic-layout-section>
            <atomic-layout-section section="main" min-width="1000px" max-width="1200px"></atomic-layout-section>
        </atomic-layout>`);
      expect(result).toMatchSnapshot();
    });

    it(`when the atomic-layout component contains a status section which contains a atomic-refine-toggle component
    should match snapshot`, async () => {
      const result =
        await getSearchLayout(`<atomic-layout id="abc" layout="search">
            <atomic-layout-section section="status">
                <atomic-refine-toggle></atomic-refine-toggle>
            </atomic-layout-section>
        </atomic-layout>`);
      expect(result).toMatchSnapshot();
    });
  });
});
