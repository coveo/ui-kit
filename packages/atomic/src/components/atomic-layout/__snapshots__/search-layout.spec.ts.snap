// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Layout #buildSearchLayout when the atomic-search-layout component contains a main & facets sections
    should match snapshot 1`] = `
"atomic-search-layout#abc { display: grid }@media only screen and (min-width: 1024px) {
      atomic-search-layout#abc {
        grid-template-areas:
          '. .                     atomic-section-search .'
          '. atomic-section-facets atomic-section-main   .'
          '. atomic-section-facets .                     .';
        grid-template-columns: 
          1fr minmax(200px, 300px) minmax(1000px, 1200px) 1fr;
        column-gap: var(--atomic-layout-spacing-x);
      }

      atomic-search-layout#abc atomic-layout-section[section=\\"facets\\"] {
        display: block;
      }
    }"
`;

exports[`Search Layout #buildSearchLayout when the atomic-search-layout component contains a status section which contains a atomic-refine-toggle component
    should match snapshot 1`] = `
"atomic-search-layout#abc { display: grid }atomic-search-layout#abc atomic-layout-section[section=\\"status\\"] atomic-sort-dropdown {
      display:none;
    }

    @media only screen and (min-width: 1024px) {
     atomic-search-layout#abc atomic-layout-section[section=\\"status\\"] atomic-sort-dropdown {
       display:block;
      }

      atomic-search-layout#abc atomic-layout-section[section=\\"status\\"] atomic-refine-toggle {
        display: none;
       }
    }"
`;

exports[`Search Layout #buildSearchLayout when the atomic-search-layout component is empty
    should match snapshot 1`] = `"atomic-search-layout#abc { display: grid }"`;
